### PR TITLE
AIP-84 Refactor Filter Query Parameters

### PR DESCRIPTION
**Pull request by jason810496**
related: #43407

Hi @pierrejeambrun,

As we discussed in https://github.com/apache/airflow/pull/43407#pullrequestreview-2405200579, here is the current implementation and usage for the refactored filter query parameters. I'd appreciate your feedback before proceeding with refactoring other endpoints to align with this approach.

Since parameter types need to be specified statically to generate the OpenAPI schema, it is necessary to have a corresponding `filter_param_factory` for each common type ( Like `float_range_filter_factory` and `datetime_range_filter_factory` for `RangeFilter` ).

Thanks!

### CODE REVIEW

File: airflow/api_fastapi/common/parameters.py [hunk 1]
```diff
@@ -239,6 +240,146 @@ def inner(order_by: str = default or self.get_primary_key_string()) -> SortParam
...
+        filter_option: FilterOptionEnum = FilterOptionEnum.EQUAL,
+        skip_none: bool = True,
+    ) -> None:
+        super().__init__(value, skip_none)
+        self.attribute: ColumnElement = attribute
+        self.value: T | None = value
+        self.filter_option: FilterOptionEnum = filter_option
+
+    def to_orm(self, select: Select) -> Select:
+        if self.value is None and self.skip_none:
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/common/parameters.py]: When you do the actually implementation, careful, some filters do not allow to set `skip_none` to False, and raise an exception.

File: airflow/api_fastapi/common/parameters.py [hunk 2]
```diff
@@ -239,6 +240,146 @@ def inner(order_by: str = default or self.get_primary_key_string()) -> SortParam
...
+            return select
+
+        if isinstance(self.value, list):
+            if self.filter_option == FilterOptionEnum.IN:
+                return select.where(self.attribute.in_(self.value))
+            if self.filter_option == FilterOptionEnum.NOT_IN:
+                return select.where(self.attribute.notin_(self.value))
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/common/parameters.py:256]: I think for tags, pool, etc..., there something a bit more complicated with 
```
        conditions = [DagModel.tags.any(DagTag.name == tag) for tag in self.value]
        return select.where(or_(*conditions))
```

That is equivalent to a generic `ArrayContains`

File: airflow/api_fastapi/common/parameters.py [hunk 3]
```diff
@@ -239,6 +240,146 @@ def inner(order_by: str = default or self.get_primary_key_string()) -> SortParam
...
+        if self.filter_option == FilterOptionEnum.LESS_THAN_EQUAL:
+            return select.where(self.attribute <= self.value)
+        if self.filter_option == FilterOptionEnum.GREATER_THAN:
+            return select.where(self.attribute > self.value)
+        if self.filter_option == FilterOptionEnum.GREATER_THAN_EQUAL:
+            return select.where(self.attribute >= self.value)
+        raise ValueError(f"Invalid filter option {self.filter_option} for value {self.value}")
+
+    def depends(self, *args: Any, **kwargs: Any) -> Self:
+        raise NotImplementedError("Use filter_param_factory instead , depends is not implemented.")
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/common/parameters.py:279]: When we remove everything from the old system, we can remove that as well.

File: airflow/api_fastapi/core_api/routes/public/event_logs.py [hunk 4]
```diff
@@ -90,46 +96,47 @@ def get_event_logs(
...
-    included_events: list[str] | None = Query(None),
-    before: datetime | None = None,
-    after: datetime | None = None,
+    dag_id: Annotated[FilterParam[str | None], Depends(str_filter_param_factory(Log.dag_id))],
+    task_id: Annotated[FilterParam[str | None], Depends(str_filter_param_factory(Log.task_id))],
+    run_id: Annotated[FilterParam[str | None], Depends(str_filter_param_factory(Log.run_id))],
+    map_index: Annotated[FilterParam[int | None], Depends(int_filter_param_factory(Log.map_index))],
+    try_number: Annotated[FilterParam[int | None], Depends(int_filter_param_factory(Log.try_number))],
+    owner: Annotated[FilterParam[str | None], Depends(str_filter_param_factory(Log.owner))],
+    event: Annotated[FilterParam[str | None], Depends(str_filter_param_factory(Log.event))],
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/core_api/routes/public/event_logs.py]: This we need to think about. It's a little bit verbous and not super easy to write. Maybe we can have only one meta `factory` that will call the appropriate `type_factory` based on the given type ?

Maybe we can just extract that into the `parameterx.py` file, so we can just reuse some:
```
LogTaskIdFilter = Annotated[FilterParam[str | None], Depends(str_filter_param_factory(Log.dag_id))]
```

To not have to copy past that line between routes. Not sure, but we can try to make this easier for the developer, that would be great.

**Review by pierrejeambrun — COMMENTED** [review]
Definitely worth pursuing, thanks.

Just a bunch of early comments.

**Comment by pierrejeambrun** [comment]
I didn't run it myself but we need to make sure that the documentation is well detected and constructed. On `http://localhost:29091/docs` (default values, types, parameter names, schemas etc..)

**Comment by jason810496** [comment]
Hi @pierrejeambrun, 

What are your thoughts on the current `filter_param_factory` implementation? The current approach adds a new parameter, `_type`, to specify the type of query parameter at runtime. This implementation consolidates multiple factories like `str_filter_param_factory`, `str_list_filter_param_factory`, etc., into a single `filter_param_factory`.

I have checked the OpenAPI schema at `http://localhost:29091/docs`, and the correct type is displayed in the documentation. 

Looking forward to your feedback!

**Comment by pierrejeambrun** [comment]
I think it's great, last time I tried exactly the same, code was working but documentation was missing the appropriate type. I think I was missing the following to make it work.
```
depends_filter.__annotations__["value"] = _type
```

But indeed having a Generic factory is the best approach, removing need to type specific factories.

Love it.

**Comment by jason810496** [comment]
Hi @pierrejeambrun,
I chose the schema with only basic functionalities defined in `common/parameters.py` but used solely in a single endpoint.
The refactor will not affect the OpenAPI schema or test cases. The current differences in `v1-generated.yaml` are caused by the `Offset`, `Limit`, and `OrderBy` parameters.

### Follow-Up
I noticed that `_SearchParam` could leverage the factory pattern as well. I plan to refactor this part after this PR is merged to avoid making the scope of this PR too large.

### Question
For some parameter schemas like `QueryTIStateFilter`, which are only used twice in `task_instance.py`, should such parameter schemas be placed directly in `task_instance.py` rather than in `common/parameters`?

**Comment by pierrejeambrun** [comment]
> For some parameter schemas like QueryTIStateFilter, which are only used twice in task_instance.py, should such parameter schemas be placed directly in task_instance.py rather than in common/parameters?

As long as they are 'reused' and not specific to a fonction I think it's fine to factorize them into `parameters.py`

**Review by pierrejeambrun — COMMENTED** [review]
Nice, should we generalize that to all routes ?

**Comment by jason810496** [comment]
> Nice, should we generalize that to all routes ?

Yes, I believe most routers can adopt `filter_param_factory` if there isn't complex logic within `to_orm`.

**Review by pierrejeambrun — COMMENTED** [review]
I think we can do the full refactoring and apply that to all route.

I'm not super comfortable having both old and new parameters system in place and live together until we decide to finalize the refactoring.

A lot of endpoints are impacted, but the change is quite limited, spec and tests are not modified. I would say the change is not that big. (And that would confirm that every case can be handled with that new system)

**Comment by jason810496** [comment]
Hi @pierrejeambrun, the refactor for `_SearchParam` is complete. Overall, only these 4 parameters remain unchanged as they don't contain common filter logic:

- `_TagsFilter`
- `_OwnersFilter`
- `_DagIdAssetReferenceFilter`

Let me know if further adjustments are needed.

**Comment by jason810496** [comment]
Rebased to the latest `main`. Looking forward to further feedback before merging 🙌

**Review by pierrejeambrun — APPROVED** [review]
Really nice.

Also as I understand, we keep flexibility by either:
- using `filter_param_factory` for `classic` orm operation that are already handled
- can still subclass `FilterParam` and implemet a full custom filter if needed.

**Comment by pierrejeambrun** [comment]
Rebasing, merging on green CI.

**Pull request merged**